### PR TITLE
Strip leading zeros from ports

### PR DESCRIPTION
**Pull request by venthur**
fixed parsing for ports with more than 5 ascii digits (i.e. lots of leading zeros). I've changed the regexp to strip leading zeros from the port in a non-capturing group.

Also added a test.

closes: #2804

### CODE REVIEW

File: src/urllib3/util/url.py [hunk 1]
```diff
@@ -61,7 +61,7 @@
 _BRACELESS_IPV6_ADDRZ_RE = re.compile("^" + _IPV6_ADDRZ_PAT[2:-2] + "$")
 _ZONE_ID_RE = re.compile("(" + _ZONE_ID_PAT + r")\]$")
 
-_HOST_PORT_PAT = ("^(%s|%s|%s)(?::([0-9]{0,5}))?$") % (
+_HOST_PORT_PAT = ("^(%s|%s|%s)(?::([0-9]*))?$") % (
```
Comment (Ousret) [inline src/urllib3/util/url.py:64]: Bit of a warning about this. We have to restrict the size.

```
int("0"*10000 + "1")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: Exceeds the limit (4300) for integer string conversion: value has 10001 digits
```

Comment (Ousret) [reply src/urllib3/util/url.py:64]: Maybe simply removing leading zeros would do it. Nonetheless, I am not sure about what is reasonable, len-wise. <4300 for sure.

Comment (venthur) [reply src/urllib3/util/url.py:64]: I see, I think we can fix the regex to catch any leading zeros in the port in a non-capturing group, discarding them.

**Review by sethmlarson — APPROVED** [review]
LGTM

**Comment by pquentin** [comment]
This pull requested broke the test suite of elastic-transport-python: https://github.com/elastic/elastic-transport-python/pull/97.

Old behavior:

```
>>> from urllib3.util import parse_url
>>> parse_url("https://[::1]:0")
Url(scheme='https', auth=None, host='[::1]', port=0, path=None, query=None, fragment=None)
```

New behavior:

```
>>> from urllib3.util import parse_url
>>> parse_url("https://[::1]:0")
Url(scheme='https', auth=None, host='[::1]', port=None, path=None, query=None, fragment=None)
>>> 
```

The port 0 is reserved but valid, I don't think we should return None? If we consider the interpretation from #2806 valid, we should make that zero part of the group. The `rfc3986` library does returns None, but `urllib.parse.urlparse` from the standard library returns 0, which seems more correct.

@sigmavirus24 Any opinions here?

**Comment by venthur** [comment]
Interesting problem! The way I read the RFC, leading zeros are permitted in the port part, and there's the "normalization" i.e. the port can be removed when empty (or in sync with the scheme) -- that seems aligned with our current implementation.

On the other hand: I cannot find a mention of a single zero in the port in the RFC. @pquentin you mention it is reserved, can you point me to the part where that is mentioned?

**Comment by pquentin** [comment]
It's not in RFC 3986, but here: https://www.iana.org/assignments/service-names-port-numbers/service-names-port-numbers.xhtml. Which I guess is my point, it should not be treated specially.

**Comment by sethmlarson** [comment]
@venthur The wording of "can be removed when empty" means if there's no data within the port section (ie `http://localhost:`) not a zero value. I believe the new behavior of setting the `port` value to `None` is a bug.

**Comment by venthur** [comment]
I'm trying to find some clever regex that removes all zeroes but one, but i stand defeated!

**Comment by sethmlarson** [comment]
@venthur  How about something like: `0*?(|0|[1-9][0-9]{0,4})`

**Comment by venthur** [comment]
thanks @sethmlarson that seems to work, I've created a pr: https://github.com/urllib3/urllib3/pull/2849

**Pull request merged**